File: src/main/java/redis/embedded/model/Shard.java
package redis.embedded.model;

import redis.embedded.core.PortProvider;

import java.util.LinkedList;
import java.util.List;

public final class Shard {

    public final String name;
    public final int mainNodePort;
    public final List<Integer> replicaPorts = new LinkedList<>();

    public Shard(final String name, int replicaCount, final PortProvider provider) {
        this.name = name;
        this.mainNodePort = provider.get();
        while (replicaCount-- > 0) {
            this.replicaPorts.add(provider.get());
        }
    }

}

File: src/main/kotlin/io/github/tobi/laa/embedded/valkey/cluster/sharded/Shard.kt
package io.github.tobi.laa.embedded.valkey.cluster.sharded

import redis.embedded.core.PortProvider
import java.util.stream.IntStream.range
import kotlin.streams.toList

data class Shard(val name: String, val mainNodePort: Int, val replicaPorts: List<Int>) {

    constructor(name: String, provider: PortProvider, replicaCount: Int) : this(
        name,
        provider.get(),
        range(0, replicaCount).map { provider.get() }.toList()
    )
}
